File: benchmarks/livermore_canonical/k24_find_min.f
C     Kernel 24 -- Find Location of First Minimum (canonical)
C     From "The Livermore Fortran Kernels" (UCRL-53745), F.H. McMahon, 1986
C     Canonical body:
C       X(n/2) = -1.0d+10
C       m = 1
C       DO 24 k=2,n: IF (X(k) .LT. X(m)) m=k
C
      PROGRAM K24
      IMPLICIT DOUBLE PRECISION (A-H,O-Z)
      INTEGER REP, M
      DIMENSION X(1001)
C
C     Initialize X using SIGNEL
C
      FUZZ = 1.2345D-3
      BUZZ = 1.0D0 + FUZZ
      FIZZ = 1.1D0 * FUZZ
      DO 10 K = 1, 1001
         BUZZ = (1.0D0 - FUZZ) * BUZZ + FUZZ
         FUZZ = -FUZZ
         X(K) = (BUZZ - FIZZ) * 0.1D0
   10 CONTINUE
C
      N = 1001
      M = 1
C
      DO 500 REP = 1, 40750000
         X(N/2) = -1.0D+10
         M = 1
         DO 200 K = 2, N
            IF (X(K) .LT. X(M)) M = K
  200    CONTINUE
  500 CONTINUE
C
      WRITE(*,901) M
      WRITE(*,902) X(M)
  901 FORMAT('m = ', I6)
  902 FORMAT('x(m) = ', E25.15)
      END

File: benchmarks/livermore_canonical/k24_find_min.c
/* K24 -- Find Location of First Minimum (canonical, 1-based indexing)
   Canonical body:
     X(n/2) = -1.0d+10
     m = 1
     DO 24 k=2,n: IF (X(k) .LT. X(m)) m=k */
#include <stdio.h>
#include <time.h>
#include "signel.h"

#define N     1001
#define NREPS 30600000

int main(void) {
    double x[1002];

    signel(x+1, 1001);

    struct timespec t0, t1;
    clock_gettime(CLOCK_MONOTONIC, &t0);

    long m = 1;
    for (long rep = 0; rep < NREPS; rep++) {
        x[N/2] = -1.0e+10;
        m = 1;
        for (long k = 2; k <= N; k++) {
            if (x[k] < x[m]) m = k;
        }
    }

    clock_gettime(CLOCK_MONOTONIC, &t1);
    double elapsed = (t1.tv_sec - t0.tv_sec) + (t1.tv_nsec - t0.tv_nsec) * 1e-9;
    printf("elapsed: %.6f s\n", elapsed);
    printf("m = %ld\n", m);
    printf("x(m) = %f\n", x[m]);
    return 0;
}
